### PR TITLE
Add ePBS stuff to consensus-types: block

### PR DESCRIPTION
**Pull request by terencechain**
This PR introduces ePBS objects into the `consensus-types/blocks` package. See the following three sections: Key changes, Existing bug fixes, and Notes to the reviewers.

## Key Changes
- Integrated ePBS block type into `NewSignedBeaconBlock`, `NewBeaconBlock`, and `NewBeaconBlockBody`, with proper error handling for blind blocks that do not apply to ePBS.
- Added getters `SignedExecutionPayloadHeader()` and `PayloadAttestations()` to the block interface.
- Updated methods including `Copy()`, `PbGenericBlock()`, `HashTreeRoot()`, `IsBlinded()`, and SSZ-related block getters to accommodate ePBS blocks.
- Updated `Proto` methods to support ePBS block handling.
- Added block setters with `SetExecutionPayloadHeader` and `SetPayloadAttestations`.
- Included `BeaconBlockePBS` in the `SignRequest` protobuf.
- Added copy methods for signed beacon block, payload attestation, and execution payload header. This is tested in the main `Copy()` method
- ~Added `executionPayloadEpbs` to satisfy execution data interface~

## Existing Bug Fix
- Corrected `SignedExecutionPayloadHeader`'s message field from `ExecutionPayloadHeader` to `ExecutionPayloadHeaderEPBS`.

## Notes to the Reviewer
- ~I did not modify the execution data interface, I feel like users will add that when the time comes, I want to avoid things we don't know if we will need. An inclusion list summary will likely be needed, so someone will have to pick that up.~
- All copy-related functions have been tested within the consensus type package, thus separate tests have not been added. I am open to adding them upon request.

### CODE REVIEW

File: consensus-types/blocks/getters.go [hunk 1]
```diff
@@ -1086,6 +1187,30 @@ func (b *BeaconBlockBody) BlobKzgCommitments() ([][]byte, error) {
...
+		return b.payloadAttestations, nil
+	default:
+		return nil, errIncorrectBlockVersion
+	}
+}
+
+// SignedExecutionPayloadHeader returns the signed execution payload header in the block.
+func (b *BeaconBlockBody) SignedExecutionPayloadHeader() (*enginev1.SignedExecutionPayloadHeader, error) {
+	switch b.version {
+	case version.Phase0, version.Altair, version.Bellatrix, version.Capella, version.Deneb:
```
Comment (potuz) [inline consensus-types/blocks/getters.go:1205]: Same here

File: consensus-types/blocks/proto.go [hunk 2]
```diff
@@ -923,6 +998,29 @@ func initBlockBodyFromProtoDeneb(pb *eth.BeaconBlockBodyDeneb) (*BeaconBlockBody
...
+		graffiti:                     bytesutil.ToBytes32(pb.Graffiti),
+		proposerSlashings:            pb.ProposerSlashings,
+		attesterSlashings:            pb.AttesterSlashings,
+		attestations:                 pb.Attestations,
+		deposits:                     pb.Deposits,
+		voluntaryExits:               pb.VoluntaryExits,
+		syncAggregate:                pb.SyncAggregate,
+		blsToExecutionChanges:        pb.BlsToExecutionChanges,
+		payloadAttestations:          pb.PayloadAttestations,
+		signedExecutionPayloadHeader: pb.SignedExecutionPayloadHeader,
```
Comment (potuz) [inline consensus-types/blocks/proto.go:1019]: Any reason to switch the order of PTC attestations with the payload header wrt `Proto()`?

File: consensus-types/blocks/setters.go [hunk 3]
```diff
@@ -99,7 +100,7 @@ func (b *SignedBeaconBlock) SetSyncAggregate(s *eth.SyncAggregate) error {
 // SetExecution sets the execution payload of the block body.
 // This function is not thread safe, it is only used during block creation.
 func (b *SignedBeaconBlock) SetExecution(e interfaces.ExecutionData) error {
-	if b.version == version.Phase0 || b.version == version.Altair {
+	if b.version == version.Phase0 || b.version == version.Altair || b.version == version.EPBS {
```
Comment (potuz) [inline consensus-types/blocks/setters.go:103]:
```suggestion
	if b.version >= version.EPBS || b.version  < version.Bellatrix {
```

File: consensus-types/blocks/proto.go (hunk 2 above)
Comment (terencechain) [reply consensus-types/blocks/proto.go:1019]: No reason and sharp eyes!

File: consensus-types/blocks/getters.go [hunk 4]
```diff
@@ -229,9 +236,21 @@ func (b *SignedBeaconBlock) PbBlindedDenebBlock() (*eth.SignedBlindedBeaconBlock
 	return pb.(*eth.SignedBlindedBeaconBlockDeneb), nil
 }
 
+// PbEPBSBlock returns the underlying protobuf object.
+func (b *SignedBeaconBlock) PbEPBSBlock() (*eth.SignedBeaconBlockEpbs, error) {
+	if b.version != version.EPBS || b.IsBlinded() {
```
Comment (potuz) [inline consensus-types/blocks/getters.go:241]: still a check for blinded

File: consensus-types/blocks/getters.go [hunk 5]
```diff
@@ -336,9 +355,12 @@ func (b *SignedBeaconBlock) ToBlinded() (interfaces.ReadOnlySignedBeaconBlock, e
 }
 
 func (b *SignedBeaconBlock) Unblind(e interfaces.ExecutionData) error {
-	if e.IsNil() {
+	if e == nil || e.IsNil() {
 		return errors.New("cannot unblind with nil execution data")
 	}
+	if b.version >= version.EPBS {
+		return consensus_types.ErrNotSupported("Unblind", b.version)
+	}
```
Comment (potuz) [inline consensus-types/blocks/getters.go:363]: Why this extra check if anyway we will fail below?

File: consensus-types/blocks/setters.go [hunk 6]
```diff
@@ -132,3 +133,21 @@ func (b *SignedBeaconBlock) SetBlobKzgCommitments(c [][]byte) error {
 		return errIncorrectBlockVersion
 	}
 }
+
+// SetPayloadAttestations sets the payload attestations in the block.
+func (b *SignedBeaconBlock) SetPayloadAttestations(p []*eth.PayloadAttestation) error {
+	if b.version < version.EPBS {
+		return consensus_types.ErrNotSupported("PayloadAttestations", b.version)
+	}
+	b.block.body.payloadAttestations = p
```
Comment (potuz) [inline consensus-types/blocks/setters.go:142]: Do we need to check the length of `p` here? or this is enforced in the pipeline.

File: proto/prysm/v1alpha1/BUILD.bazel [hunk 7]
```diff
@@ -65,8 +65,8 @@ ssz_gen_marshal(
         "BeaconBlockBodyCapella",
         "SignedBeaconBlockCapella",
         "BeaconBlockDeneb",
-	"BeaconBlockePBS",
-	"SignedBeaconBlockePBS",
+	    "BeaconBlockEpbs",
+	    "SignedBeaconBlockEpbs",
```
Comment (potuz) [inline proto/prysm/v1alpha1/BUILD.bazel:69]: better to remove the extra indentation, Preston typically complains about this

File: consensus-types/blocks/setters.go (hunk 6 above)
Comment (terencechain) [reply consensus-types/blocks/setters.go:142]: I think it should be enforced by the caller (in the pipeline), and the function doesn't care, similar to the rest of the setters. I'm happy to add a check if you think it should be here.

File: proto/prysm/v1alpha1/BUILD.bazel (hunk 7 above)
Comment (terencechain) [reply proto/prysm/v1alpha1/BUILD.bazel:69]: weird, this actually fixes it to align with the rest, but before it was off : )
Screenshot 2024-04-24 at 11 17 02 AM

Comment (terencechain) [reply proto/prysm/v1alpha1/BUILD.bazel:69]: This is how engine build.bazel looks to me 
Screenshot 2024-04-24 at 11 18 00 AM

**Review by potuz — APPROVED** [review]

**Pull request merged**